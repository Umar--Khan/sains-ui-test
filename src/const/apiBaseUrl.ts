export const apiBaseUrl = 'https://jsainsburyplc.github.io/front-end-test/';
