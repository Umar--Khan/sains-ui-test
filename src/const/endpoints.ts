const endpoints = {
  getProducts: 'products.json',
};

export default endpoints;
