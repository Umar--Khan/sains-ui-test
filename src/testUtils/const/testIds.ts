export const headerTestIds = {
  numberInBasket: 'number-in-basket',
};

export const storeTestIds = {
  productCard: 'store-product-card',
};

export const basketTestIds = {
  reviewCard: 'basket-review-card',
  emptyBasketMessage: 'empty-basket-message',
};
