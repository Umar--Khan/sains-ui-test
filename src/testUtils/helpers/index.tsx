import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';

import { BasketContext, BasketWrapper } from 'src/context/BasketContext';

export const withRouter = (children: any) => (
  <MemoryRouter>{children}</MemoryRouter>
);

export const renderWithBasketContext = (value: BasketWrapper, children: any) =>
  render(
    <BasketContext.Provider value={value}>
      {withRouter(children)}
    </BasketContext.Provider>,
  );
