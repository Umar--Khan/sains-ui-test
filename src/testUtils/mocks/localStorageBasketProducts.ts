import { BasketProduct } from 'src/views/Store/Store.d';

export const mockedLocalStorageBasketProducts = [
  {
    productDetail: {
      'productId': '5493179',
      'sku': '549/3179',
      'title': 'Russell Hobbs Pennine Illuminating S Steel Kettle 20444',
      'price': 24.99,
      'image': 'https://media.4rgos.it/s/Argos/9805244_R_SET?w=110&h=130',
    },
    quantity: 5,
  },
  {
    productDetail: {
      'productId': '4251824',
      'sku': '425/1824',
      'title': 'Challenge White Desk Fan - 12 Inch',
      'price': 14.99,
      'image': 'https://media.4rgos.it/s/Argos/4251824_R_SET?w=110&h=130',
    },
    quantity: 1,
  },
  {
    productDetail: {
      'productId': 'tuc132404445',
      'sku': 'tuc132404445',
      'title': 'Plain Vest Top 12 BLACK',
      'price': 16,
      'image': 'https://media.4rgos.it/s/Argos/tuc132404445_R_SET?w=110&h=130',
    },
    quantity: 2,
  },
] as BasketProduct[];
