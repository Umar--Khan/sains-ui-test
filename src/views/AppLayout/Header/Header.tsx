import { Button } from 'shared/Buttons';

const Header = () => (
  <header>
    <Button>Your basket</Button>
  </header>
);

export default Header;
