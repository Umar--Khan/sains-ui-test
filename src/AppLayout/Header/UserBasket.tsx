import { headerTestIds } from 'src/testUtils/const/testIds';

interface UserBasketProps {
  itemsInBasket: number;
}

const UserBasket = ({ itemsInBasket }: UserBasketProps) => (
  <span data-testid={headerTestIds.numberInBasket}>
    Basket ({itemsInBasket} items)
  </span>
);

export default UserBasket;
