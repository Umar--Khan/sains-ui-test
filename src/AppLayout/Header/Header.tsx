import { useContext } from 'react';
import { NavLink as RouterNavLink } from 'react-router-dom';
import styled from 'styled-components';

import { ROUTES_MAP } from 'src/const/routes';
import { BasketContext } from 'src/context/BasketContext';
import * as colors from 'src/styles/_colors.module.scss';
import UserBasket from './UserBasket';

const StyledHeader = styled.header`
  background-color: ${colors.white};
  border-bottom: 2px solid ${colors.greyLight};
`;

const NavContainer = styled.nav`
  display: flex;
  justify-content: flex-end;
  align-items: center;
`;

const HeaderContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  align-items: center;
  min-height: 4rem;
  max-width: 90rem;
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  margin: auto;
  @media (min-width: 720px) {
    width: 70%;
  }
  @media (min-width: 720px) and (max-width: 1200px) {
    width: 75%;
  }
`;

const NavLink = styled(RouterNavLink)`
  margin-left: 1.25rem;
  text-decoration: none;
  font-size: 1.25rem;
  line-height: 1.5rem;
  font-weight: 700;
  color: ${colors.greyDark};

  &:hover {
    text-decoration: underline;
    color: ${colors.orangeDark};
  }

  &.active {
    text-decoration: underline;
    color: ${colors.orange};
  }
`;

const Header = () => {
  const { basketLocal } = useContext(BasketContext);

  const itemsInBasket = basketLocal?.reduce<number>((acc, { quantity }) => {
    acc += quantity;
    return acc;
  }, 0);

  return (
    <StyledHeader>
      <HeaderContainer>
        <NavContainer>
          <NavLink to={ROUTES_MAP.STORE}>Store</NavLink>
          <NavLink to={ROUTES_MAP.BASKET}>
            <UserBasket itemsInBasket={itemsInBasket ? itemsInBasket : 0} />
          </NavLink>
        </NavContainer>
      </HeaderContainer>
    </StyledHeader>
  );
};

export default Header;
