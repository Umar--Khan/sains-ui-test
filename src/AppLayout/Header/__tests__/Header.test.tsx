import renderer from 'react-test-renderer';

import { mockedLocalStorageBasketProducts } from 'src/testUtils/mocks/localStorageBasketProducts';
import { renderWithBasketContext, withRouter } from 'src/testUtils/helpers';
import { headerTestIds } from 'src/testUtils/const/testIds';
import Header from '../Header';

jest.mock('src/hooks/useResource');

describe('<Header />', () => {
  it('Should render correctly', () => {
    const component = renderer.create(withRouter(<Header />));
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Should render 0 items in basket if theyre no items in local storage', () => {
    const { getByTestId } = renderWithBasketContext(
      {
        basketLocal: [],
        setBasketLocal: jest.fn(),
      },
      <Header />,
    );

    const basketItem = getByTestId(headerTestIds.numberInBasket);
    expect(basketItem.textContent).toEqual('Basket (0 items)');
  });

  it('Should render correct items number in basket if theyre items in local storage', () => {
    const { getByTestId } = renderWithBasketContext(
      {
        basketLocal: mockedLocalStorageBasketProducts,
        setBasketLocal: jest.fn(),
      },
      <Header />,
    );

    const basketItem = getByTestId(headerTestIds.numberInBasket);
    expect(basketItem.textContent).toEqual('Basket (8 items)');
  });
});
