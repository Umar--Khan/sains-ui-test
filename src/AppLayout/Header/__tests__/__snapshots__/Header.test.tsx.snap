// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Should render correctly 1`] = `
<header
  className="sc-bdfBwQ JrDbG"
>
  <div
    className="sc-dlfnbm cNcuHH"
  >
    <nav
      className="sc-gsTCUz fHRLXO"
    >
      <a
        aria-current={null}
        className="sc-hKgILt cCXpce"
        href="/store"
        onClick={[Function]}
      >
        Store
      </a>
      <a
        aria-current={null}
        className="sc-hKgILt cCXpce"
        href="/basket"
        onClick={[Function]}
      >
        <span
          data-testid="number-in-basket"
        >
          Basket (
          0
           items)
        </span>
      </a>
    </nav>
  </div>
</header>
`;
